scylla-cdc/src/log_reader.rs

```diff
@@ -0,0 +1,213 @@
+                }
+                Ok(_) = self.end_timestamp_receiver.changed() => {
+                    let timestamp = *self.end_timestamp_receiver.borrow_and_update();
+                    self.end_timestamp = timestamp;
+                    CDCReaderWorker::set_upper_timestamp(&self.readers, timestamp).await;
+                }
+            }
+
+            if stream_reader_tasks.is_empty() {
+                if let Some(err) = err {
+                    return Err(err);
+                }
+
+                if let Some(generation) = next_generation {
+                    next_generation = None;
+
+                    if generation.timestamp > self.end_timestamp {
+                        return Ok(());
+                    }
+
+                    self.readers = fetcher
+                        .fetch_stream_ids(&generation)
+                        .await?
+                        .iter()
+                        .flatten()
+                        .map(|stream_id| {
+                            Arc::new(StreamReader::new(
+                                &self.session,
+                                vec![stream_id.clone()],
+                                max(self.start_timestamp, generation.timestamp),
+                                self.window_size,
+                                self.safety_interval,
+                                self.sleep_interval,
+                            ))
+                        })
+                        .collect();
+
+                    stream_reader_tasks = self
+                        .readers
+                        .iter()
+                        .map(|reader| {
+                            let reader = Arc::clone(reader);
+                            let keyspace = self.keyspace.clone();
+                            let table_name = self.table_name.clone();
+                            let factory = Arc::clone(&self.consumer_factory);
+                            tokio::spawn(async move {
+                                let consumer = factory.new_consumer().await;
+                                reader.fetch_cdc(keyspace, table_name, consumer).await
+                            })
+                        })
+                        .collect();
+                } else if had_first_generation {
+                    // FIXME: There may be another generation coming in the future with timestamp < end_timestamp
+                    // that could have been missed because of earlier fetching failures.
+                    // More on this here https://github.com/piodul/scylla-cdc-rust/pull/10#discussion_r826865162
+                    return Ok(());
+                }
+            }
+        }
+    }
+
+    async fn set_upper_timestamp(readers: &[Arc<StreamReader>], new_upper_timestamp: Duration) {
+        // This function does not take &self, because that would require the CDCWorker to be Send.
```

Comment:
Do you know why `CDCReaderWorker` is not `Send`? The printer worker had a similar function and it worked fine.

Reply:
It was probably a relict from the times when `ConsumerFactory` wasn't `Sync` and I forgot to change it, my bad.